realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -1146,12 +1150,20 @@ JNIEXPORT jlongArray JNICALL Java_io_realm_internal_TableQuery_nativeBatchUpdate
         std::vector<jlong> exported_handover_tableview_array(number_of_queries);
 
         // Step1: Position the shared group at the handover query version so we can import all queries
-
         // read the first query to determine the version we should use
         SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, handover_queries_pointer_array[0]);
         std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
-        // position this shared group at the specified version
-        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+
+        // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
+        // The Handover object doesn't prevent a SharedGroup version from no longer being accessible. In rare
+        // cases this means that the version in the Handover object is invalid and Realm Core will throw a
+        // BadVersion as result.
+        if (SG(bgSharedGroupPtr)->get_transact_stage() != SharedGroup::transact_Reading) {
+            SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+        } else if (SG(bgSharedGroupPtr)->get_version_of_current_transaction() != handoverQuery->version) {
```

Comment:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  jcjacksonengineering@gmail.com
    Domain joe-jackson.net has exceeded the max emails per hour (118/100 (118%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext8.iad.github.net ([192.30.252.199]:59282 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by rs5.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86_1)
	(envelope-from <noreply@github.com>)
	id 1aflk9-001g97-6F
	for jcjackson@joe-jackson.net; Tue, 15 Mar 2016 05:58:25 -0400
Date: Mon, 14 Mar 2016 10:31:57 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1457976717;
	bh=sEScRBD8HAEm/7agcSwGLXQZTtNcW0/zVoSdGQWrcV0=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=WpdL8KoPBLsseb5gqjKCm4Lb00tzRm8mQtQFbVqu9+1M5hHK8+9gn97e5oc01cPUY
	 9Jm7HAIaU/rKCnrEr23q0oSTR3kdmQy+OQNwLp20vD/N3ExatzMsK4PPP6F+JGjA5t
	 SdmLhr0fQwYE3q+6T3RIQmAbgBVKm0xvzdJhsVlY=
From: SeongTaekC <notifications@github.com>
Reply-To: realm/realm-java <reply@reply.github.com>
To: realm/realm-java <realm-java@noreply.github.com>
Message-ID: <realm/realm-java/pull/2450/r56040975@github.com>
In-Reply-To: <realm/realm-java/pull/2450@github.com>
References: <realm/realm-java/pull/2450@github.com>
Subject: Re: [realm-java] Fix BadVersionException (#2450)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_56e6f58d5fd46_6a123fcc42de729c1497eb";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: stk1m1
X-GitHub-Recipient: jcj52436999
List-ID: realm/realm-java <realm-java.realm.github.com>
List-Archive: https://github.com/realm/realm-java
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000f94f70392f0b03b64ee11d4b639cdae1a6fee27b24d6192cf0000000112feb78d92a170ce03571e0f@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AA-U90OtBErGm9MEVHiflkgtR-EfjSEoks5ptZuNgaJpZM4HwK21>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: jcjackson@joe-jackson.net


----==_mimepart_56e6f58d5fd46_6a123fcc42de729c1497eb
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

>          // read the first query to determine the version we should use
>          SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, handover_queries_pointer_array[0]);
>          std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
> -        // position this shared group at the specified version
> -        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
> +
> +        // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
> +        // The Handover object doesn't prevent a SharedGroup version from no longer being accessible. In rare
> +        // cases this means that the version in the Handover object is invalid and Realm Core will throw a
> +        // BadVersion as result.
> +        if (SG(bgSharedGroupPtr)->get_transact_stage() != SharedGroup::transact_Reading) {
> +            SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
> +        } else if (SG(bgSharedGroupPtr)->get_version_of_current_transaction() != handoverQuery->version) {

@cmelchior I am not trying to review. @finnschiermer says 
> I don't think Core promises this call to be well defined, when used outside a transaction. It should be called from within a transaction, but see below for a bigger problem.

You can see his comment in my unit test for new `distinct()` submitted [here](https://github.com/realm/realm-core/commit/078e61244aa1551b7dcf475c4fa648e1dd62585c#diff-dfb86ea2c75fcee0e86ed82cdc5ae7adR11129)


---
Reply to this email directly or view it on GitHub:
https://github.com/realm/realm-java/pull/2450/files#r56040975
----==_mimepart_56e6f58d5fd46_6a123fcc42de729c1497eb
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>In <a href="https://github.com/realm/realm-java/pull/2450#discussion_r56040975">realm/realm-jni/src/io_realm_internal_TableQuery.cpp</a>:</p>
<pre style='color:#555'>&gt;          // read the first query to determine the version we should use
&gt;          SharedGroup::Handover&lt;Query&gt; *handoverQueryPtr = HO(Query, handover_queries_pointer_array[0]);
&gt;          std::unique_ptr&lt;SharedGroup::Handover&lt;Query&gt;&gt; handoverQuery(handoverQueryPtr);
&gt; -        // position this shared group at the specified version
&gt; -        SG(bgSharedGroupPtr)-&gt;begin_read(handoverQuery-&gt;version);
&gt; +
&gt; +        // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
&gt; +        // The Handover object doesn&#39;t prevent a SharedGroup version from no longer being accessible. In rare
&gt; +        // cases this means that the version in the Handover object is invalid and Realm Core will throw a
&gt; +        // BadVersion as result.
&gt; +        if (SG(bgSharedGroupPtr)-&gt;get_transact_stage() != SharedGroup::transact_Reading) {
&gt; +            SG(bgSharedGroupPtr)-&gt;begin_read(handoverQuery-&gt;version);
&gt; +        } else if (SG(bgSharedGroupPtr)-&gt;get_version_of_current_transaction() != handoverQuery-&gt;version) {
</pre>
<p><a href="https://github.com/cmelchior" class="user-mention">@cmelchior</a> I am not trying to review. <a href="https://github.com/finnschiermer" class="user-mention">@finnschiermer</a> says </p>

<blockquote>
<p>I don't think Core promises this call to be well defined, when used outside a transaction. It should be called from within a transaction, but see below for a bigger problem.</p>
</blockquote>

<p>You can see his comment in my unit test for new <code>distinct()</code> submitted <a href="https://github.com/realm/realm-core/commit/078e61244aa1551b7dcf475c4fa648e1dd62585c#diff-dfb86ea2c75fcee0e86ed82cdc5ae7adR11129">here</a></p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/realm/realm-java/pull/2450/files#r56040975">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AA-U9y4uRe2tr_YmjN3FVvGNMvMxytCvks5ptZuNgaJpZM4HwK21.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/realm/realm-java/pull/2450/files#r56040975"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_56e6f58d5fd46_6a123fcc42de729c1497eb--